zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/CassandraTableInstaller.java

```diff
@@ -0,0 +1,261 @@
+        model
+            .getColumns().stream()
+            .map(ModelColumn::getColumnName)
+            .map(ColumnName::getStorageName)
+            .anyMatch(c -> !databaseColumns.contains(c));
+
+    return !isAnyColumnNotCreated;
+  }
+
+  public void createTable(Model model, long timeBucket) {
+    try {
+      final String table = TableHelper.getTable(model, timeBucket);
+      createOrUpdateTable(model, table, model.getColumns(), false);
+      createOrUpdateTableIndexes(model, table, model.getColumns(), false);
+      createAdditionalTable(model, timeBucket);
+    } catch (Exception e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  public void createAdditionalTable(Model model, long timeBucket) throws SQLException {
+    final Map<String, SQLDatabaseModelExtension.AdditionalTable> additionalTables = model.getSqlDBModelExtension().getAdditionalTables();
+    for (final SQLDatabaseModelExtension.AdditionalTable table : additionalTables.values()) {
+      final String tableName = TableHelper.getTable(table.getName(), timeBucket);
+      createOrUpdateTable(model, tableName, table.getColumns(), true);
+      createOrUpdateTableIndexes(model, tableName, table.getColumns(), true);
+    }
+  }
+
+  public void createOrUpdateTable(Model model, String table, List<ModelColumn> columns, boolean isAdditionalTable) {
+    try {
+      final List<ModelColumn> columnsToBeAdded = new ArrayList<>(columns);
+      final Set<String> existingColumns = getDatabaseColumns(table);
+
+      columnsToBeAdded.removeIf(it -> existingColumns.contains(it.getColumnName().getStorageName()));
+
+      final KeyspaceMetadata metadata = ((CassandraClient) this.client).getMetadata();
+      if (!metadata.getTable(table).isPresent()) {
+        createTable(model, table, columnsToBeAdded, isAdditionalTable);
+      } else {
+        updateTable(table, columnsToBeAdded);
+      }
+    } catch (Exception e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  public void createOrUpdateTableIndexes(Model model, String table, List<ModelColumn> columns, boolean isAdditionalTable) throws SQLException {
+    final CassandraClient cassandraClient = (CassandraClient) this.client;
+
+    final List<String> columnsMissingIndex =
+        columns
+            .stream()
+            .filter(ModelColumn::shouldIndex)
+            .filter(it -> it.getLength() < 256)
+            .filter(c -> !model.isRecord() || c.getBanyanDBExtension().getShardingKeyIdx() > 0)
+            .map(ModelColumn::getColumnName)
+            .map(ColumnName::getStorageName)
+            .collect(toList());
+
+    // adding the time_bucket as an index column when querying zipkin_query
+    columnsMissingIndex.add("time_bucket");
+    for (String column : columnsMissingIndex) {
+      final String index = "idx_" + table + "_" + column;
+      if (!indexExists(cassandraClient, table, index)) {
+        executeSQL(
+            new SQLBuilder("CREATE INDEX ")
```

Comment:
1. Avoid indexes, especially legacy 2i.  (v2 uses SASI, if you must better is to wait for 5.0 and use SAI, at minimum create a design that can still run without the index – true scalability as well as availability when an index needs to be rebuilt)
2. `indexExists(…)` does not work in a distributed environment.  it can be used as an optimisation but cannot be relied upon. use `IF NOT EXISTS`.  use the optimisation everywhere though (CREATE TABLE too).
3.

Reply:
> Avoid indexes, especially legacy 2i. (v2 uses SASI, if you must better is to wait for 5.0 and use SAI, at minimum create a design that can still run without the index – true scalability as well as availability when an index needs to be rebuilt)

Updated, still using the SASI index for now. I could update the SAI index when Cassandra v5.0 is ready. 

> indexExists(…) does not work in a distributed environment. it can be used as an optimisation but cannot be relied upon. use IF NOT EXISTS. use the optimisation everywhere though (CREATE TABLE too).

Update to using the pre-defined table CQL file.